### ISSUE TITLE
needless_borrow track through function call

### ISSUE THREAD

**Issue by nik-suri** [rust-lang/rust-clippy#10464]
### Summary

Currently, the `needless_borrow` check is described as:
> What it does
Checks for address of operations (&) that are going to be dereferenced immediately by the compiler.

This misses cases where we have borrows earlier in a function that aren't used. For example,
```
pub struct S;

pub fn f(_: &S) {}

pub fn g() {
    let s = &&S;
    f(s);
}

pub fn h() {
    let s = S;
    f(&&s);
}
```

In the above example, `needless_borrow` is triggered on function `h`, but not on function `g`.

### Lint Name

needless_borrow

### Reproducer

Can use same example code block given above and verify that the `needless_borrow` warning is triggered on function `h` but not on function `g`:
```
pub struct S;

pub fn f(_: &S) {}

pub fn g() {
    let s = &&S;
    f(s);
}

pub fn h() {
    let s = S;
    f(&&s);
}
```

### Version

```text
rustc 1.67.1 (d5a82bbd2 2023-02-07)
cargo 1.67.1 (8ecd4f20a 2023-01-10)

System information:
 Server Version: 20.10.18
 Storage Driver: overlay2
  Backing Filesystem: xfs
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Kernel Version: 5.15.0-1030-aws
 Operating System: Ubuntu 20.04.5 LTS
 OSType: linux
 Architecture: x86_64
```
